Core/MIPS/MIPSVFPUUtils.cpp

```diff
@@ -245,14 +246,44 @@ void ReadMatrix(float *rd, MatrixSize size, int reg) {
 	default: _assert_msg_(JIT, 0, "%s: Bad matrix size", __FUNCTION__);
 	}
 
-	for (int i = 0; i < side; i++) {
-		for (int j = 0; j < side; j++) {
-			int index = mtx * 4;
-			if (transpose)
-				index += ((row+i)&3) + ((col+j)&3)*32;
-			else
-				index += ((col+j)&3) + ((row+i)&3)*32;
-			rd[j*4 + i] = V(index);
+	// The voffset ordering is now integrated in these formulas,
+	// eliminating a table lookup.
+	const float *v = currentMIPS->v + (size_t)mtx * 16;
+	if (transpose) {
+		if (side == 4 && col == 0 && row == 0) {
+			// Simple 4x4 transpose. TODO: Optimize.
+			for (int j = 0; j < 4; j++) {
+				for (int i = 0; i < 4; i++) {
+					rd[j * 4 + i] = v[i * 4 + j];
+				}
+			}
+		} else {
+			for (int j = 0; j < side; j++) {
+				for (int i = 0; i < side; i++) {
+					int index = ((row + i) & 3) * 4 + ((col + j) & 3);
+					rd[j * 4 + i] = v[index];
+				}
+			}
+		}
+	} else {
+		// Fast path
+		if (side == 4 && col == 0 && row == 0) {
+			memcpy(rd, v, sizeof(float) * 16);
+			/*
+			for (int j = 0; j < side; j++) {
```

Comment:
Personally I think we can skip these comments or just have one line commented out i.e. `// rd[N] = v[mtx * 16 + N];`.

-[Unknown]